File: Original/app/src/main/java/com/andruid/magic/newsdaily/adapter/CountryAdapter.java
package com.andruid.magic.newsdaily.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.andruid.magic.newsdaily.databinding.LayoutCountryBinding;
import com.andruid.magic.newsdaily.viewholder.CountryViewHolder;
import com.blongho.country_data.Country;

import java.util.List;

public class CountryAdapter extends RecyclerView.Adapter<CountryViewHolder> {
    final private List<Country> countries;

    public CountryAdapter(List<Country> countries){
        this.countries = countries;
    }

    @NonNull
    @Override
    public CountryViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        LayoutCountryBinding binding = LayoutCountryBinding.inflate(inflater, parent, false);
        return new CountryViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull CountryViewHolder holder, int position) {
        Country country = countries.get(position);
        holder.bindCountry(country);
    }

    @Override
    public int getItemCount() {
        return countries.size();
    }
}

File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/adapter/CountryAdapter.kt
package com.andruid.magic.newsdaily.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import com.andruid.magic.newsdaily.databinding.LayoutCountryBinding
import com.andruid.magic.newsdaily.viewholder.CountryViewHolder
import com.blongho.country_data.Country

class CountryAdapter : ListAdapter<Country, CountryViewHolder>(CountryDiffCallback()) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CountryViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = LayoutCountryBinding.inflate(inflater, parent, false)
        return CountryViewHolder(binding)
    }

    override fun onBindViewHolder(holder: CountryViewHolder, position: Int) {
        val country: Country = getItem(position)
        holder.bind(country)
    }

    class CountryDiffCallback : DiffUtil.ItemCallback<Country>() {
        override fun areItemsTheSame(oldItem: Country, newItem: Country): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: Country, newItem: Country): Boolean {
            return oldItem.id == newItem.id && oldItem.name == newItem.name &&
                    oldItem.flagResource == newItem.flagResource
        }

    }
}